File: engine/src/com/magma/engine/ui/dialog/Dialog.java
package com.magma.engine.ui.dialog;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input.Keys;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.scenes.scene2d.ui.Stack;
import com.badlogic.gdx.scenes.scene2d.ui.TextArea;
import com.badlogic.gdx.utils.Queue;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.magma.engine.assets.MagmaLoader;
import com.magma.engine.assets.Shapes;

import space.earlygrey.shapedrawer.ShapeDrawer;

public class Dialog extends Stack {

    private static Dialog instance;
    private final Queue<DialogMessage> messages;

    protected ShapeDrawer shapes;
    private TextArea text;

    private boolean isLocked;

    public Dialog(Viewport view, int width, int height) {
        Dialog.instance = this;
        this.shapes = Shapes.getInstance();
        this.messages = new Queue<DialogMessage>();
        this.text = new TextArea("_", MagmaLoader.getDebugSkin());
        this.text.setDisabled(true);
        setPosition((view.getWorldWidth() - width) * 0.5f, -view.getWorldHeight() + 10);
        setSize(width, height);
        Gdx.app.log("Dialog",
                "Dialog created at bounds: " + getX() + " , " + getY() + " (" + getWidth() + " x " + getHeight() + ")");
        addActor(text);
    }

    @Override
    public void act(float delta) {
        if (messages.isEmpty()) {
            return;
        }

        if (Gdx.input.isKeyJustPressed(Keys.ENTER)) {
            if (!isLocked) {
                messages.removeFirst();
            } else {
                Gdx.app.log("Dialog", "Dialog is locked");
                return;
            }
        }

        toFront();
        super.act(delta);
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        if (messages.isEmpty()) {
            return;
        }

        DialogMessage message = messages.first();
        if (message.style == DialogStyle.Basic) {
            shapes.setColor(Color.BLACK);
            shapes.filledRectangle(getX(), getY(), getWidth(), getHeight());
            shapes.setColor(Color.WHITE);
            shapes.rectangle(getX(), getY(), getWidth(), getHeight(), 3f);
            text.setText(message.message);
        } else {
            throw new IllegalArgumentException("No dialog style defined!");
        }
        super.draw(batch, parentAlpha);
    }

    public static boolean isSpeaking() {
        return !getInstance().messages.isEmpty();
    }

    public static void speak(DialogMessage message) {
        Gdx.app.log(">", message.message);
        getInstance().messages.addLast(message);
    }

    public static void speak(String message) {
        speak(new DialogMessage(message));
    }

    public static void speak(int key) {
        speak(DialogMessage.PLACEHOLDER);
    }

    public static boolean isLocked() {
        return getInstance().isLocked;
    }

    public static void lock() {
        getInstance().isLocked = true;
    }

    public static void unlock() {
        getInstance().isLocked = false;
    }

    public static Dialog getInstance() {
        if (instance == null) {
            throw new NullPointerException("No UI to show dialog! Create at least one Dialog instance!");
        }
        return instance;
    }
}

File: engine/src/com/magma/engine/ui/dialog/Dialog.kt
package com.magma.engine.ui.dialog

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.scenes.scene2d.ui.Stack
import com.badlogic.gdx.scenes.scene2d.ui.TextArea
import com.badlogic.gdx.utils.Queue
import com.magma.engine.assets.MagmaLoader
import com.magma.engine.assets.Shapes
import com.magma.engine.debug.MagmaLogger
import com.magma.engine.stages.ViewportContext

class Dialog(width: Int, height: Int) : Stack() {
    private val text: TextArea = TextArea("_", MagmaLoader.debugSkin)

    init {
        x = (ViewportContext.width - width) * 0.5f
        y = -ViewportContext.width + 10
        setSize(width.toFloat(), height.toFloat())
        Gdx.app.log(
            "Dialog",
            "Dialog created at bounds: " + x + " , " + y + " (" + getWidth() + " x " + getHeight() + ")"
        )
        addActor(text)
    }

    override fun act(delta: Float) {
        if (messages.isEmpty) {
            return
        }
        if (Gdx.input.isKeyJustPressed(Input.Keys.ENTER)) {
            if (!isLocked) {
                messages.removeFirst()
            } else {
                Gdx.app.log("Dialog", "Dialog is locked")
                return
            }
        }
        toFront()
        super.act(delta)
    }

    override fun draw(batch: Batch, parentAlpha: Float) {
        if (messages.isEmpty) {
            return
        }
        val message = messages.first()
        val shapes = Shapes.instance
        if (message.style == DialogStyle.Basic) {
            shapes.setColor(Color.BLACK)
            shapes.filledRectangle(x, y, width, height)
            shapes.setColor(Color.WHITE)
            shapes.rectangle(x, y, width, height, 3f)
            text.text = message.message
        } else {
            throw IllegalArgumentException("No dialog style defined!")
        }
        super.draw(batch, parentAlpha)
    }

    companion object {
        private val messages: Queue<DialogMessage> = Queue()
        private var isLocked = false
        var locked: Boolean
            get() = isLocked
            set(value) {if (value) MagmaLogger.log(this,"Dialog locked") else MagmaLogger.log(this,"Dialog unlocked")field =
                isLocked = value
            }

        val isSpeaking: Boolean
            get() = messages.isEmpty

        fun speak(message: DialogMessage) {
            Gdx.app.log(">", message.message)
            messages.addLast(message)
        }

        fun speak(message: String) {
            speak(DialogMessage(message))
        }

        fun speak(key: Int) {
            speak(DialogMessage.PLACEHOLDER)
        }
    }
}